### PR TITLE
Preset

### PR DESCRIPTION
**Pull request by dyuvzhenko**
New user interface

### CODE REVIEW

File: demo/package.json [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  "version": "0.1.0",
+  "private": true,
+  "dependencies": {
+    "react": "^16.6.3",
+    "react-dom": "^16.6.3",
+    "react-scripts": "2.1.1",
+    "wikitext-markdown": "file:../wikitext-markdown-0.0.0.tgz"
+  },
+  "scripts": {
+    "start": "cd .. && npm pack && cd demo && npm install ../wikitext-markdown-0.0.0.tgz && react-scripts start",
```
Comment (dhilt) [inline demo/package.json]: 👍

File: demo/src/App.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      wikiText: this.converter.toWiki(e.target.value)
+    })
+  }
+
+  render() {
+    const { wikiText, markdownText } = this.state
+    return (
+      <div style={{'textAlign': 'center'}}>
+        <h2>Convert data from wiki to markdown</h2>
+        <textarea style={textareaStyle} value={wikiText} onChange={this.changeWikiText.bind(this)} />
```
Comment (dhilt) [inline demo/src/App.js]: my eyes are bleeding, bind in render after all...

File: demo/src/App.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        <textarea style={textareaStyle} value={markdownText} onChange={this.changeMarkdownText.bind(this)} />
```
Comment (dhilt) [inline demo/src/App.js]: again... `changeMarkdownText = (e) => { ... }` instead?

File: package.json [hunk 4]
```diff
@@ -1,8 +1,8 @@
 {
   "name": "wikitext-markdown",
   "version": "0.0.0",
-  "description": "",
-  "main": "index.js",
+  "description": "Package for convert data from wiki-markup to markdown-markup, and back",
```
Comment (dhilt) [inline package.json]: "A library for converting wiki to markdown and back"

File: src/index.js [hunk 5]
```diff
@@ -1,4 +1,62 @@
...
+  toWiki(text) {
+    let newText = text.split('\n')
+    this.toWikiRules.forEach(rule => newText = newText.map(rule))
+    return newText.join('\n')
+  }
+
+  toMarkdown(text) {
+    let newText = text.split('\n')
+    this.toMarkdownRules.forEach(rule => newText = newText.map(rule))
+    newText = newText.filter(line => line !== 'BLOCKQUOTE_START' && line !== 'BLOCKQUOTE_END') // TODO: remove that
```
Comment (dhilt) [inline src/index.js]: todo or not todo?
btw it looks hacky...

File: src/index.js [hunk 6]
```diff
@@ -1,4 +1,62 @@
...
+class Converter {
+  constructor(settings = {}) {
+    this.preset = settings.preset || 'default'
+    /* init rules */
+    let toMarkdownRules = getRulesForParseWikitextToMarkdown(this.preset)
+    let toWikiRules = getRulesForParseMarkdownToWikitext(this.preset)
+
+    /* skip rules selected by user */
+    if (settings.skip && Array.isArray(settings.skip)) {
+      if (settings.skip.some(e => e === 'blockquote')) { // TODO: remove that
```
Comment (dhilt) [inline src/index.js]: so, remove that
or remove todo

File: src/preset.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
+const defaultParseToMarkdown = require('./parse-rules/default/wikitext-to-markdown');
+const defaultParseToWiki = require('./parse-rules/default/markdown-to-wikitext');
+
+const htmlParseToMarkdown = require('./parse-rules/html/wikitext-to-markdown');
+const htmlParseToWiki = require('./parse-rules/html/markdown-to-wikitext');
+
+function getRulesForParseWikitextToMarkdown(preset) {
```
Comment (dhilt) [inline src/preset.js:15]: need to have a white list of presets and turn this procedure into more strict form:
```
presets = [ 'presetA', 'presetB', 'default-with-html', ... ]
...
(preset) => {
  if (presets.indexOf(preset) === -1) {
    console.warn(`Invalid preset name ${preset}. Fallback to default`);
  }
  switch (preset) {
    case 'default-with-html':
      return [...defaultParseToMarkdown, ...htmlParseToMarkdown];
    default:
      return defaultParseToMarkdown;
  }
}
```

File: src/parse-rules/default/markdown-to-wikitext.js [hunk 8]
```diff
@@ -51,54 +51,29 @@ function parseBlockquoteSecondPass(line, index, arr) {
...
-    .map(options.skipCode && skip || parseCode)
-    .map(options.skipBlockquote && skip || parseBlockquoteFirstPass)
-    .map(options.skipBlockquote && skip || parseBlockquoteSecondPass)
-    .map(options.skipLink && skip || parseLink)
-    .map(options.skipBoldAndItalic && skip || parseBoldAndItalic)
-    .map(options.skipBoldAndItalic && skip || parseBold)
-    .map(options.skipBoldAndItalic && skip || parseItalic)
-    .join('\n')
-}
+module.exports = [
```
Comment (dhilt) [inline src/parse-rules/default/markdown-to-wikitext.js:68]: I would prefer to have an object with methods here:

```
markdownToWikitextRules = {
  ['headings']: (line) => { 
    ... 
  },

  ...

  ['italic']: (line) => { 
    ... 
  }
}

module.exports = markdownToWikitextRules;
```

So rules tokens are explicit now and could be extracted into constant enumeration and reused:

```
RULES.ITALIC = 'italic';
...
  [RULES.ITALIC]: (line) => { 
    ... 
  }
```

File: src/parse-rules/html/wikitext-to-markdown.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+module.exports = [
+  // htmlOrderedList,
+  // htmlSetSequenceForOrderedList,
+  // htmlHeadings,
+  // htmlUnorderedList,
+  // htmlCode,
+  // htmlBoldAndItalic,
+  // htmlBlockquote,
+  // htmlRemoveSpecialSymbols,
+  // htmlLink,
```
Comment (dhilt) [inline src/parse-rules/html/wikitext-to-markdown.js]: do we really want all these stub rules? why not to remove them?

File: package.json [hunk 10]
```diff
@@ -1,8 +1,8 @@
...
+  "main": "src/index.js",
   "scripts": {
     "test": "jest"
   },
```
Comment (dhilt) [inline package.json:8]: need also "author", "keywords" fields to be present in package.json

**Review by dhilt — CHANGES_REQUESTED** [review]

File: src/preset.js [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+  return preset === 'default' ? defaultParseToMarkdown :
+    preset === 'default-with-html' ? [...defaultParseToMarkdown, ...htmlParseToMarkdown] :
+      [] /* no rules for wrong preset name */
+}
+
+function getRulesForParseMarkdownToWikitext(preset) {
```
Comment (dhilt) [inline src/preset.js:25]: the same note as for `getRulesForParseWikitextToMarkdown`

**Review by dyuvzhenko — COMMENTED** [review]
Comment

File: README.md [hunk 12]
```diff
@@ -0,0 +1,62 @@
+# A npm library for converting wiki to markdown and back ![Build status](https://travis-ci.com/khajjit/wikitext-markdown.svg?branch=master)
+
+### Install
+
+`npm install wikitext-markdown`
+
+### Usage
+
```
Comment (khajjit) [inline README.md]: Comment 2

**Pull request closed without merging**